File: basicAlgorithms/_binarySearch.py
import math

def binarySearch (arr,target):

    l = 0
    r = len(arr) - 1

    while l<=r:
        m = l + math.floor((r-l)/2)

        if target == arr[m]:
            return m
        elif target < arr[m]:
            r = m - 1
        elif target > arr[m]:
            l = m + 1
    
    return -1

print(binarySearch([-2,-1,0,1,2,3,4,5,6,7,8,9,10],4))

File: basicAlgorithms/_binarySearch.cpp
#include <iostream>
#include <vector>
#include <cmath>

int _binarySearch(std::vector<int>& nums, int target){
    int l = 0;
    int r = nums.size() - 1;
    int m = 0;

    while(l<=r){
        m = l + floor((r-l)/2);

        if(target == nums[m]){
            return m;
        }
        else if(target < nums[m]){
            r = m - 1;
        }
        else if(target > nums[m]){
            l = l + 1;
        }
    }
    
    return -1;
}

int main(){
    std::vector arr = {-2,-1,0,1,2,3,4,5,6,7,8,9,10};
    int result = _binarySearch(arr,4);
    std::cout << result << std::endl;
    return 0;
}

